File: Problems/238.product-of-array-except-self.cpp
// title: Product of Array Except Self
// timestamp: 2024-11-01 11:53:44
// problemUrl: https://leetcode.com/problems/product-of-array-except-self/
// memory: 40.3 MB
// runtime: 2 ms

class Solution {
public:
    vector<int> productExceptSelf(vector<int>& nums) {
        vector<int> res(nums.size()), r(nums.size(), 1), l(nums.size(), 1);
        for (int i = 1; i < nums.size(); i++)
            l[i] = l[i - 1] * nums[i - 1];
        for (int i = nums.size() - 2; i >= 0; i--)
            r[i] = r[i + 1] * nums[i + 1];
        for (int i = 0; i < nums.size(); i++)
            res[i] = l[i] * r[i];
        
        return res;
    }
};

File: Problems/238.product-of-array-except-self.go
// title: Product of Array Except Self
// timestamp: 2024-04-18 15:18:35
// problemUrl: https://leetcode.com/problems/product-of-array-except-self/
// memory: 8 MB
// runtime: 32 ms

func productExceptSelf(nums []int) []int {
    leftProd, rightProd := make([]int, len(nums)), make([]int, len(nums))
    rightProd[0], leftProd[len(nums) - 1] = 1, 1
    for i := 1; i < len(nums) - 1; i++ {
        rightProd[i] = rightProd[i - 1] * nums[i - 1]
        leftProd[len(nums) - i - 1] = leftProd[len(nums) - i] * nums[len(nums) - i]
    }
    
    rightProd[len(nums) - 1] = rightProd[len(nums) - 2] * nums[len(nums) - 2]
    leftProd[0] = leftProd[1] * nums[1]

    for i := 0; i < len(leftProd); i++ {
        leftProd[i] *= rightProd[i]
    }

    return leftProd
}